packages/api-client/src/granules.ts

```diff
@@ -30,6 +30,7 @@ type AssociateExecutionRequest = {
 export const getGranuleResponse = async (params: {
   prefix: string,
   granuleId: GranuleId,
+  duplicateHandling?: string,
```

Comment:
Of course, a 404 could eventually succeed, but retries on 404s should be handled explicitly, as the [waitForGranule](https://github.com/nasa/cumulus/blob/8136e220e0db03922bf0ad530760352ae4c758bd/packages/api-client/src/granules.ts#L95) function does.

Reply:
So - I'll take a closer look at this PR in a bit, but *generally* I don't think we should be adding non-generic handling code into the api-client to deal with task concerns.     This package's purpose is to simply pass along the payload to the PrivateApi invocation and handle the expected status codes/retries.

Reply:
Looking at the ticket, the block

```
  try {
    response = await granules.getGranuleResponse({
      prefix: process.env.STACKNAME,
      granuleId,
    });
  } catch (error) {
    const responseError = error;
    if (responseError.statusCode === 404) {
      return granuleId;
    }
    throw new Error(`Unexpected error from Private API lambda: ${responseError.message}`);
  }
  ```
  
  seems to pretty clearly attempt to catch a 404 - which is fine, except the API client has default retires. 
  
  Is there a reason we don't pass in a default pRetry setup for this call using  `pRetryOptions` to set the retries to 1,  or override the onFailedAtempt method to abort the pRetry in case of 404?

Reply:
Or better yet, just have 404 be an `expectedStatusCode`?

Reply:
Do we really want to do that? I don't think we want to set retries to 1, because while that works nicely for 404s, it also means that other errors won't be retried either, which is not what we want, do we? This is why I suggest making use of `AbortError` for 4xx errors in `invokeApi` so that 4xx errors are not retried, but will still cause pRetry to reject.

Looking at `waitForGranule`, I think my suggested approach also would fix what appears to be an unintended explosion of retries there. Since `waitForGranule` invokes `getGranuleResponse`, as the code currently stands, a 404 could potentially be retried 30 times because `getGranuleResponse` retries up to 3 times (via `invokeApi`) and `waitForGranule` retries `getGranuleResponse` up to 10 times.

By changing `invokeApi` so that it does not retry 4xx errors (and throws `AbortError` instead, to signal to `pRetry` to abort without retries for 4xx responses), `getGranuleResponse` will thus not trigger undesired 404 retries, and thus `waitForGranule` will make only up to 10 retries instead of up to 30.

Did I read the code correctly?

Reply:
`I'm not sure why we don't pass in pRetryOptions here, I was looking through the repo and it seemed like pRetryOptions is only ever actually used for testing/ helper functions for testing. `

Sure, but those  tests are clients of this module.       Generally exposing the retry logic so long as it has sensible defaults seems like a good, not bad, thing, and I'd argue it's probably a bad thing we haven't exposed those for more of the methods.          The invoke method is deliberately generic.

```
Looking at this issue further, I believe the correct solution is to leave this code alone (as you suggest), and instead, modify the [invokeApi](https://github.com/nasa/cumulus/blob/63e83812660db89b06923821c1885f88e4e4b5ea/packages/api-client/src/cumulusApiClient.ts#L25) function to properly abort retrying when there is a 404 response (actually any 4xx response since they are all client errors, and thus retries will never succeed), as shown in the [pRetry usage example](https://github.com/sindresorhus/p-retry#usage), where AbortError is used.
```

I think this approach is a misapplication of the intent of the inovkeApi method in that package- making the *default* onFailed attempt care about status codes makes it non-generic, and is a misapplication in cases where we're *expecting* to get a 404, calls like: 

```
  await t.notThrowsAsync(providersApi.deleteProvider({
    prefix: t.context.testPrefix,
    providerId: t.context.testProviderId,
    expectedStatusCodes: [404, 200],
    callback,
  }));
```

We *want*  to support that sort of thing, and applying it to the generic method is, in my opinion, not the right approach.     A more maintainable approach for the client package would be to expose the options inherent in invokeApi that already support what we're trying to do.   This has two benefits: 

1)   Doesn't make a breaking change to this exported method
2)   Allows a modification like the following (off the cufff for illustration only) that *doesn't* rely on error trapping an expected response: 


```
const checkGranuleHasNoDuplicate = async (granuleId, duplicateHandling) => {
  let response;
  try {
    response = await granules.getGranuleResponse({
      prefix: process.env.STACKNAME,
      granuleId,
      expectedStatusCodes: [404, 200]
    });
  } catch (error) {
    throw new Error(`Unexpected error from Private API lambda: ${error.message}`);
  }
  
  if (response.statusCode === 404) {
      return granuleId;
  }

  if (response.statusCode === 200) {
    if (duplicateHandling === 'error') {
      throw new Error(`Duplicate granule found for ${granuleId} with duplicate configuration set to error`);
    }
    return false;
  }

  throw new Error(`Unexpected return from Private API lambda: ${JSON.stringify(response)}`);
};
```

Reply:
I agree with Jonathan's approach, in this case we don't want to retry on [404, 200]

Reply:
Sounds good to me! Thanks for talking that through @Jkovarik!